File: python/IDGenerator.py
class IDGenerator:

    current = 1
    
    ids = {}    
    maxID = 100
    for index in range (maxID):
        ids[index+1] = False    

    @staticmethod
    def _next():
        
        if(IDGenerator.current == -1):
            for index in range(1, IDGenerator.maxID+1):
                if(not IDGenerator.ids[index]):
                    
                    IDGenerator.current = index
                    return True
        else:
            for index in range(IDGenerator.current, IDGenerator.maxID):
                
                if(not IDGenerator.ids[index+1]):                
                    
                    IDGenerator.current = index+1
                    return True
            for index in range(1, IDGenerator.current):
                
                if(not IDGenerator.ids[index]):
                    IDGenerator.current = index
                    return True
            IDGenerator.current = -1
            return False
    
    @staticmethod
    def GetNext():
         
        if(IDGenerator.current!=-1):
            newId = IDGenerator.current
            IDGenerator.ids[newId] = True;
            
            IDGenerator._next()
            
            return newId
        else:
            IDGenerator._next()
            if(IDGenerator.current == -1):
                return -1
            newId = IDGenerator.current
            IDGenerator.ids[newId] = True;
            IDGenerator._next()
            return newId
            #IDGenerator.MoveToNextFreeID()

        
    @staticmethod
    def HasID():
        for index in range(1, IDGenerator.maxID+1):
            if(not IDGenerator.ids[index]):
                return True
        
        return False
    
    @staticmethod
    def DeleteByID(GenID):
        if(IDGenerator.ids[GenID] == True):
            IDGenerator.ids[GenID] = False
            return True
        else:
            #print "the id is free"
            return False

    @staticmethod
    def IncreaseLimitBy(additionIdNumbers):
        for index in range(IDGenerator.maxID, IDGenerator.maxID + additionIdNumbers):
            IDGenerator.ids[index+1] = False

        IDGenerator.maxID = IDGenerator.maxID + additionIdNumbers
        
    @staticmethod        
    def TestGenerator1():
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "TakeId():",IDGenerator.GetNext()
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "TakeId():",IDGenerator.GetNext()
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "TakeId():",IDGenerator.GetNext()
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "ReleaseID(3)", IDGenerator.DeleteByID(3)
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"
        print "MoveToNextFreeID()", IDGenerator.GetNext()
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"    
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "ReleaseID(2)", IDGenerator.DeleteByID(2)
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "TakeId():",IDGenerator.GetNext()
        print "-----------------------"
        print "IDs", IDGenerator.ids
        print "-----------------------"
        print "current : ",IDGenerator.current
        print "-----------------------"
        print "HasFreeID()",IDGenerator.HasID()
        IDGenerator.IncreaseLimitBy(5)


#IDGenerator.TestGenerator1()
        
        
        

File: python/IDGenerator.ts
class IDGenerator {
  private static current: number = 1;
  private static ids: { [key: number]: boolean } = {};
  private static maxID: number = 100;

  // Initialize IDs
  static {
    for (let index = 0; index < IDGenerator.maxID; index++) {
      IDGenerator.ids[index + 1] = false;
    }
  }

  private static _next(): boolean {
    if (IDGenerator.current === -1) {
      for (let index = 1; index <= IDGenerator.maxID; index++) {
        if (!IDGenerator.ids[index]) {
          IDGenerator.current = index;
          return true;
        }
      }
    } else {
      for (let index = IDGenerator.current; index < IDGenerator.maxID; index++) {
        if (!IDGenerator.ids[index + 1]) {
          IDGenerator.current = index + 1;
          return true;
        }
      }
      for (let index = 1; index < IDGenerator.current; index++) {
        if (!IDGenerator.ids[index]) {
          IDGenerator.current = index;
          return true;
        }
      }
      IDGenerator.current = -1;
      return false;
    }
    return false;
  }

  public static GetNext(): number {
    if (IDGenerator.current !== -1) {
      const newId = IDGenerator.current;
      IDGenerator.ids[newId] = true;
      IDGenerator._next();
      return newId;
    } else {
      IDGenerator._next();
      if (IDGenerator.current === -1) {
        return -1;
      }
      const newId = IDGenerator.current;
      IDGenerator.ids[newId] = true;
      IDGenerator._next();
      return newId;
    }
  }

  public static HasID(): boolean {
    for (let index = 1; index <= IDGenerator.maxID; index++) {
      if (!IDGenerator.ids[index]) {
        return true;
      }
    }
    return false;
  }

  public static DeleteByID(GenID: number): boolean {
    if (IDGenerator.ids[GenID]) {
      IDGenerator.ids[GenID] = false;
      return true;
    } else {
      return false;
    }
  }

  public static IncreaseLimitBy(additionIdNumbers: number): void {
    for (let index = IDGenerator.maxID; index < IDGenerator.maxID + additionIdNumbers; index++) {
      IDGenerator.ids[index + 1] = false;
    }
    IDGenerator.maxID += additionIdNumbers;
  }

  public static TestGenerator1(): void {
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("TakeId():", IDGenerator.GetNext());
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("TakeId():", IDGenerator.GetNext());
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("TakeId():", IDGenerator.GetNext());
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("ReleaseID(3)", IDGenerator.DeleteByID(3));
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("MoveToNextFreeID()", IDGenerator.GetNext());
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("ReleaseID(2)", IDGenerator.DeleteByID(2));
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("TakeId():", IDGenerator.GetNext());
    console.log("-----------------------");
    console.log("IDs", IDGenerator.ids);
    console.log("-----------------------");
    console.log("current : ", IDGenerator.current);
    console.log("-----------------------");
    console.log("HasFreeID()", IDGenerator.HasID());
    IDGenerator.IncreaseLimitBy(5);
  }
}

// Uncomment to run the test
// IDGenerator.TestGenerator1();
